stdlib/internal/promql/join.md

```diff
@@ -0,0 +1,91 @@
+### join
+
+`join` merges two input streams into a single output stream.
+Records that have the same group key and `_time` values will be joined in the output.
+The output groups will be the same as the input groups.
+
+`join` has the following properties:
+
+| Name    | Type     | Description |
+| ------- | -------- | ----------- |
+| tables  | object   | Specifies the two table streams to join. |
+| fn      | function | Defines the function that joins the rows of each table. The return value is an object which defines the output record structure. |
+
+Ex.
+```
+import "internal/promql"
+
+left = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "A")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "a"})
+
+right = from(bucket: "my-bucket")
+    |> range(start: -1h)
+    |> filter(fn: (r) => r._field == "B")
+    |> drop(columns: ["_field"])
+    |> rename(columns: {_value: "b"})
+
+promql.join(tables: {left, right}, fn: (t) => {
+    return {t.left with b: t.right.b}
```

Comment:
If there are common column labels for both tables, what is the behavior?  Based on how `with` behaves within `map`, I guess that columns on the RHS of `with` will overwrite any columns with the same name on the LHS.

Reply:
Reading this it occurs to me again that having data that is "pivoted" out of the gate (i.e., no `_field` column and `_value` renamed to be the name of the field) would simplify things a lot.

Reply:
Yes in this case they would be overwritten.

Reply:
This seems more complex than I would like and would impact the current performance because functions are slow to execute in the current system.

I'm not aware of any system that requires a function to specify how the data is joined and I don't think we should start. It is far too complex and will likely make it too difficult of a function to use in my opinion.

Additionally, I see that there ends up being a bit of a fundamental problem with joining data where you don't know all of the data. If there's any conflicts for column names, you end up with data overwritten.

I think that we should only have an `on` keyword which specifies what fields we join on (such as `_time` for most use cases), we automatically join on group keys, and any other columns get prefixed the same as the current join.

Reply:
This join is much more selective than the current flux join. Therefore in the majority of cases, we will be performing relatively few function calls. 

In my opinion, join is similar to map. You could consider it a map over two streams. I really find the implicit column renaming of the current join to be very awkward. I'd rather have the user declare exactly what they want out.

And I don't want to go too off topic, but type inference can actually catch naming conflicts at compile time in this form.

Reply:
In order to keep the scope limited, I guess the main concern I have is how would the transpiler generate the function call? If the transpiler would always generate the same function using `with`, then I don't see a purpose in the `fn` parameter.

I also am not aware of any prior work that treats join as similar to map. It will also be harder to make join efficient if it includes an implicit call to `map` and it also means that we will have another place where `map` can cause issues. What happens if I do this?

```
join(tables: {left, right}, fn: (l, r) => {host: "server{l._value}"})
```

And `host` ends up being part of the group key. It is much simpler if we declare that join will just always prefix field key names. I think the existing join behavior does some weird hybrid behavior where it only does the prefix if there's a conflict. I think the prefix should always happen.

Reply:
I'm not sure I follow your example. We would have the same issue if a user put a similar map call after join.

I've given examples of how to transpile PromQL using this join and yes they would always generate the same type of function call. But putting this join in an internal package gives us the ability to try it without breaking the public API at a later point.

And in terms of efficiency, we can add fast internal implementations which the planner can rewrite to for common use cases, so I don't see that being an issue. Even without specialized implementations, the selective nature of groupKeyJoin makes it so that we won't be evaluating the anonymous function for very many rows in most cases.